lib/web_ui/lib/src/engine/bitmap_canvas.dart

```diff
@@ -178,7 +178,11 @@ class BitmapCanvas extends EngineCanvas {
     _canvasPool.clear();
     final int len = _children.length;
     for (int i = 0; i < len; i++) {
-      _children[i].remove();
+      html.Element child = _children[i];
+      // Don't remove children that have been reused by CrossFrameCache.
+      if (child.parent == rootElement) {
```

Comment:
Is there a way to do this without reading back from the DOM? This might be an indicator of an issue in the design. It means we lost some information and now are trying to recover it. This can be brittle. For example, changing how we nest our elements affects this logic. It touches elements that are not related to caching, e.g. clipping elements. I imagine `child.parent` will also trigger allocation of `dart:html` element wrapper corresponding to the parent?

Reply:
Yes as discussed before, we should move to a model where canvas elements and images are reused at lower levels instead of recycling/matching BitmapCanvas(s).

Regarding cost, the dart2js code is fine:
```for (t2 = _this.rootElement, i = 0; i < len; ++i) {
        child = t1[i];
        if (child.parentElement === t2) {
          t3 = child.parentNode;
          if (t3 != null)
            t3.removeChild(child);
        }
      }
```